src/ipc/ipc-common/IPCProxyBase.h

```diff
@@ -43,23 +42,132 @@
+{
+    bool m_ready{};
+    std::vector<IObserver *> m_observers;
+public:
+    Monitor()=default;
+    virtual ~Monitor()=default;
+
+    void add(IObserver& observer){
+        m_observers.push_back(&observer);
+    }
+    void del(IObserver& observer){
+        m_observers.erase( remove(m_observers.begin(), m_observers.end(), &observer), m_observers.end() );
+    }
+    void notify(Monitor& source ){
+        for (auto observer: m_observers){
+            observer->handle_event(source);
+        }
+    }
+
+    bool stateReady() const { return m_ready; }
+    void stateReadyChanged(bool val)
+    {
+        this->m_ready = val;
+        notify(*this);
+    }
+};
+// IsReadyObserver
+class IsReadyObserver : public QObject
+{
+    Q_OBJECT
+    bool m_ready{};
+public:
+    IsReadyObserver(QObject *parent = 0) : QObject(parent) {};
+    Q_SIGNAL void isReadyChanged();
+
+    bool ready(){ return m_ready; }
+    bool set_ready(bool val){
+        return m_ready=val;
+    }
+    void raiseSignal(){
+        emit isReadyChanged();
+    }
+
+    template<typename Function>
+    void callReady(const QObject* context, Function f)
+    {
+        if (ready()) {
+            f();
+        }
+        else {
+            auto connection = std::make_unique<QMetaObject::Connection>();
+            *connection = QObject::connect(this, &IsReadyObserver::isReadyChanged,
+                                           context, [this, f{std::move(f)}, connection{std::move(connection)} ]()
+                {
+                    if (this->ready()) {
+                        f();
+                        QObject::disconnect(*connection);
+                    }
+                }
+            );
+        }
+    }
+};
+// Single-time observer which will unregister itself when done
+class SingleTimeObserver : public IsReadyObserver, public IObserver
+{
+    const QObject* m_context{};
+    std::function<void()> m_func{};
+public:
+    explicit SingleTimeObserver(const QObject* context, std::function<void()> func=nullptr): m_context{context}, m_func{func} {}
+    ~SingleTimeObserver()=default;
+
+    void handle_event(Monitor &source ) override {
+        if(set_ready(source.stateReady())){
+            callReady(m_context, m_func);
+            raiseSignal();
+            source.del(*this);
+        }
+    }
+};
+// Standard observer which will work for each signal
+class StandartObserver : public IsReadyObserver, public IObserver
 {
+    const QObject* m_context{};
+    std::function<void()> m_func{};
+public:
+    explicit StandartObserver(const QObject* context, std::function<void()> func=nullptr): m_context{context}, m_func{func} {}
+    ~StandartObserver()=default;
+
+    void handle_event(Monitor &source ) override {
+        if(set_ready(source.stateReady())){
+            callReady(m_context, m_func);
+            raiseSignal();
+        }
+    }
+};
 
+template<typename AdapterType>
+class IPCProxyBase : public AdapterType, Monitor, protected IPCProxyBaseBase
```

Comment:
You should not extend class further. Inheritance is the class extension. Please create member object that holds observers.